filesystems/blob/src/test/scala/cromwell/filesystems/blob/BlobPathBuilderFactorySpec.scala

```diff
@@ -1,28 +1,189 @@
+    val expiry = BlobFileSystemManager.parseTokenExpiry(sasToken)
+    expiry should contain(expiryTime)
+  }
+
+  it should "verify an unexpired token will be processed as unexpired" in {
+    val expiryTime = generateTokenExpiration(11L)
+    val expired = BlobFileSystemManager.hasTokenExpired(expiryTime, Duration.ofMinutes(10L))
+    expired shouldBe false
+  }
+
+  it should "test an expired token will be processed as expired" in {
+    val expiryTime = generateTokenExpiration(9L)
+    val expired = BlobFileSystemManager.hasTokenExpired(expiryTime, Duration.ofMinutes(10L))
+    expired shouldBe true
+  }
+
+  it should "test that a filesystem gets closed correctly" in {
+    val endpoint = BlobPathBuilderSpec.buildEndpoint("storageAccount")
+    val azureUri = BlobFileSystemManager.uri(endpoint)
+    val fileSystems = mock[FileSystemAPI]
+    val fileSystem = mock[FileSystem]
+    when(fileSystems.getFileSystem(azureUri)).thenReturn(Try(fileSystem))
+    when(fileSystems.closeFileSystem(azureUri)).thenCallRealMethod()
+
+    fileSystems.closeFileSystem(azureUri)
+    verify(fileSystem, times(1)).close()
+  }
+
+  it should "test retrieveFileSystem with expired filesystem" in {
+    val endpoint = BlobPathBuilderSpec.buildEndpoint("storageAccount")
+    val expiredToken = generateTokenExpiration(9L)
+    val refreshedToken = generateTokenExpiration(69L)
+    val sasToken = BlobPathBuilderFactorySpec.buildExampleSasToken(refreshedToken)
+    val container = BlobContainerName("storageContainer")
+    val configMap = BlobFileSystemManager.buildConfigMap(sasToken, container)
+    val azureUri = BlobFileSystemManager.uri(endpoint)
+
+    val fileSystems = mock[FileSystemAPI]
+    val blobTokenGenerator = mock[BlobTokenGenerator]
+    when(blobTokenGenerator.generateAccessToken).thenReturn(Try(sasToken))
+
+    val fsm = BlobFileSystemManager(container, endpoint, 10L, blobTokenGenerator, fileSystems, Some(expiredToken))
+    fsm.getExpiry should contain(expiredToken)
+    fsm.isTokenExpired shouldBe true
+    fsm.retrieveFilesystem()
+
+    fsm.getExpiry should contain(refreshedToken)
+    fsm.isTokenExpired shouldBe false
+    verify(fileSystems, never()).getFileSystem(azureUri)
+    verify(fileSystems, times(1)).newFileSystem(azureUri, configMap)
+    verify(fileSystems, times(1)).closeFileSystem(azureUri)
+  }
+
+  it should "test retrieveFileSystem with an unexpired fileSystem" in {
+    val endpoint = BlobPathBuilderSpec.buildEndpoint("storageAccount")
+    val initialToken = generateTokenExpiration(11L)
+    val refreshedToken = generateTokenExpiration(71L)
+    val sasToken = BlobPathBuilderFactorySpec.buildExampleSasToken(refreshedToken)
+    val container = BlobContainerName("storageContainer")
+    val configMap = BlobFileSystemManager.buildConfigMap(sasToken, container)
+    val azureUri = BlobFileSystemManager.uri(endpoint)
+    // Need a fake filesystem to supply the getFileSystem simulated try
+    val dummyFileSystem = null
```

Comment:
Is there a way to do this without a `null`?

Reply:
So yes, but it would involve another mock. I was using this way in favor of a mock mostly because we do not actually need the object, but the type signature, but I was mostly trying to use this as an exercise to only include the mocks we need to test against.